free/src/main/scala/cats/free/InvariantCoyoneda.scala

```diff
@@ -0,0 +1,108 @@
+package cats
+package free
+
+/**
+ * The free invariant functor on `F`. This is isomorphic to `F` as long as `F` itself is a
+ * invariant functor. The function from `F[A]` to `InvariantCoyoneda[F,A]` exists even when
+ * `F` is not an invariant functor. Implemented using a List of functions for stack-safety.
+ */
+sealed abstract class InvariantCoyoneda[F[_], A] extends Serializable { self =>
+  import InvariantCoyoneda.{unsafeApply, Aux}
+
+  /**
+   * The pivot between `fi` and `k`, usually existential.
+   */
+  type Pivot
+
+  /**
+   * The underlying value.
+   */
+  val fi: F[Pivot]
+
+  /**
+   * The list of the `Pivot => *` halves of the transformer functions, to be composed and lifted into `F` by `run`.
+   */
+  private[cats] val ks0: List[Any => Any]
+
+  /**
+   * The list of the `* => Pivot` halves of the transformer functions, to be composed and lifted into `F` by `run`.
+   */
+  private[cats] val ks1: List[Any => Any]
+
+  /**
+   * Half of the composed transformer function, to be lifted into `F` by `run`.
+   */
+  final def k0: Pivot => A = Function.chain(ks0.reverse)(_).asInstanceOf[A]
+
+  /**
+   * The other half of the composed transformer function, to be lifted into `F` by `run`.
+   */
+  final def k1: A => Pivot = Function.chain(ks1)(_).asInstanceOf[Pivot]
+
+  /**
+   * Converts to `F[A]` given that `F` is a invariant functor
+   */
+  final def run(implicit F: Invariant[F]): F[A] = F.imap(fi)(k0)(k1)
+
+  /**
+   * Converts to `G[A]` given that `G` is a invariant functor
+   */
+  final def foldMap[G[_]](trans: F ~> G)(implicit G: Invariant[G]): G[A] =
+    G.imap(trans(fi))(k0)(k1)
+
+  /**
+   * Simple function composition. Allows imap fusion without touching the underlying `F`.
+   */
+  final def imap[B](f: A => B)(g: B => A): Aux[F, B, Pivot] =
+    unsafeApply(fi)(f.asInstanceOf[Any => Any] :: ks0)(g.asInstanceOf[Any => Any] :: ks1)
+
+  /**
+   * Modify the context `F` using transformation `f`.
+   */
+  final def mapK[G[_]](f: F ~> G): Aux[G, A, Pivot] =
+    unsafeApply(f(fi))(ks0)(ks1)
+
+}
+
+object InvariantCoyoneda {
+
+  /**
+   * Lift the `Pivot` type member to a parameter. It is usually more convenient to use `Aux` than
+   * a refinment type.
+   */
+  type Aux[F[_], A, B] = InvariantCoyoneda[F, A] { type Pivot = B }
+
+  /**
+   * `F[A]` converts to `InvariantCoyoneda[F,A]` for any `F`
+   */
+  def lift[F[_], A](fa: F[A]): InvariantCoyoneda[F, A] =
```

Comment:
Why not preserve the Aux type here but do so in other places?

Reply:
Actually I don't see the reason to ever preserve the Aux type. I can't see how anyone can use it. Isn't it better to use the type that hides the Pivot?

Reply:
I'm just following the style of the other `Coyoneda`s, happy to make this change.

Reply:
I think it is worth it if we have AndThen designed explicitly for this purpose but it is probably binary incompatible on the existing ones since you will change List to AndThen.

Reply:
If we are going to retain the Aux in other places I think this should be `Aux[F, A, A]`.

Reply:
Makes sense. I made this change to all the Coyonedas.